### PR TITLE
[WIP] (feat) Added Exponential BackOff for Retry in builder

### PR DESCRIPTION
**Pull request by rahulchheda**
Signed-off-by: Rahul M Chheda <rahul.chheda@mayadata.io>

Fix for issue: https://github.com/fission/fission/issues/1585

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1614)

### CODE REVIEW

File: pkg/buildermgr/pkgwatcher.go [hunk 1]
```diff
@@ -100,8 +100,13 @@ func (pkgw *packageWatcher) build(buildCache *cache.Cache, srcpkg *fv1.Package)
 		return
 	}
 
+	// Create a new BackOff for health check on environment builder pod
+	healthCheckBackOff, err := utils.NewBackOff(500*time.Millisecond,time.Minute/10,5)
+	if err != nil {
+		pkgw.logger.Error("Unable to create BackOff for Health Check", zap.Error(err))
```
Comment (vishal-biyani) [inline pkg/buildermgr/pkgwatcher.go:106]:
```suggestion
		pkgw.logger.Error("Unable to create BackOff for Health Check", zap.Error(err), zap.String("package_name", pkg.ObjectMeta.Name))
```

File: pkg/buildermgr/pkgwatcher.go [hunk 2]
```diff
@@ -100,8 +100,13 @@ func (pkgw *packageWatcher) build(buildCache *cache.Cache, srcpkg *fv1.Package)
 		return
 	}
 
+	// Create a new BackOff for health check on environment builder pod
+	healthCheckBackOff, err := utils.NewBackOff(500*time.Millisecond,time.Minute/10,5)
```
Comment (vishal-biyani) [inline pkg/buildermgr/pkgwatcher.go:104]: So at least for builder pod especially for platforms like JVM the `maxInterval` of 6 seconds seems small. Can we go for something like 120 seconds?

**Comment by vishal-biyani** [comment]
Thanks for the PR @rahulchheda - except two minor comments, looks good. Also one nitpick - backoff seems like a single word so `backoff.go` might be a better name instead of camel case one. I will test this sometime today/tomorrow once you have made changes and then I will merge it

**Comment by rahulchheda** [comment]
Thanks for the review @vishal-biyani , let me just do some more modification

**Comment by rahulchheda** [comment]
Btw, @vishal-biyani , Do you want me to include some logic around maxRetries, similar to maxInterval?

**Comment by vishal-biyani** [comment]
@rahulchheda maxretries would be indeed good. I don't think trying till infinity is useful.

**Comment by rahulchheda** [comment]
Right now it checks till the currentBackOffDuration <= maxInterval, I can include one more variable to take in the MaxRetry, let me update the PR. I think this would be useable further down the line.

**Comment by rahulchheda** [comment]
Sorry for unwanted changes, will fix in final commit, PTAL @vishal-biyani

File: examples/go/module-example/go.mod [hunk 3]
```diff
@@ -1,3 +1,5 @@
 module github.com/fission/fission/examples/go/go-module-example
 
 require github.com/golang/example v0.0.0-20170904185048-46695d81d1fa
+
+go 1.13
```
Comment (vishal-biyani) [inline examples/go/module-example/go.mod:5]: Should we move this change out of this PR?

Comment (rahulchheda) [reply examples/go/module-example/go.mod:5]: Sure, got this added into git staged my mistake, will fix in next commit.

File: pkg/utils/backoff.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	DefaultInitialInterval = 500 * time.Millisecond
+	DefaultMultiplier      = 1.5
+	DefaultMaxInterval     = 60 * time.Second
```
Comment (vishal-biyani) [inline pkg/utils/backoff.go:11]: Just tested and even with 60 seconds it does not work with JVM env. I would suggest bumping up default max to 300s to cover potential other scenarios.

**Comment by rahulchheda** [comment]
@vishal-biyani Made the appropriate changes, could you take a look?

File: pkg/utils/backoff.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	DefaultInitialInterval = 500 * time.Millisecond
+	DefaultMultiplier      = 1.5
+	DefaultMaxInterval     = 300 * time.Second
+	DefaultMaxCount        = 30
```
Comment (vishal-biyani) [inline pkg/utils/backoff.go:12]: I just tested this and 30 count is happening before the max interval. Can we bump this to be inline with new max interval please?

**Comment by vishal-biyani** [comment]
@rahulchheda Thanks and sorry for delay. Can you please fix the CI build issue and one minor comment - otherwise looks good!

**Comment by vishal-biyani** [comment]
Thanks @rahulchheda this looks good - I will do a quick test and merge this later this week.

**Comment by vishal-biyani** [comment]
Thanks a lot, @rahulchheda - I verified and this works nicely!

**Pull request merged**